### PR TITLE
Updated Docker integration for local development

### PR DESCRIPTION
**Pull request by Taronyuu**
What has been done:
1. Move all Docker related files (with the exception of `docker-compose.yml`) to a separate directory to keep it more clean.
2. Updated the `docker-compose.yml` file to use the new Dockerfile
3. Modified the Dockerfile so it will build the current repository *without* having to download a release from Github. The repository itself is enough. It also solves the issue with the missing `BUDGET_VERSION` environment variable.
4. Modify the `entrypoint.sh` script so it runs the required commands to start the container when developing.

By modifying the `entrypoint.sh` script and the `Dockerfile` we make sure that the resulting container does have the generated code already, but will still re-compile when being used for local development. (Since the local directory is only mounted after the container has been build/started)

### CODE REVIEW

**Comment by range-of-motion** [comment]
Thanks! A few generic questions.

* Wouldn't it make sense to move `entrypoint.sh` into the `docker/` directory as well?
* I agree with updating the MySQL version from 5.7 to 8, just not sure if we should include it in this scope

File: docker-compose.yml [hunk 1]
```diff
@@ -14,8 +25,5 @@ services:
     ports:
       - 3306:3306
     command: --default-authentication-plugin=mysql_native_password
-
-  app:
-    build: ./
-    ports:
-      - 8080:80
+    volumes:
+      - db-data:/var/lib/mysql:rw
```
Comment (range-of-motion) [inline docker-compose.yml:29]: Why is this necessary exactly?

File: entrypoint.sh [hunk 2]
```diff
@@ -1,5 +1,16 @@
...
+su - www-data << 'EOF'
+    composer install --no-dev --no-interaction
+
+    npm install
+    npm run build
+
+    cp .env.docker .env
+
+    php artisan storage:link
+EOF
```
Comment (range-of-motion) [inline entrypoint.sh:12]: What's up with this? We're already executing operations like these in `Dockerfile` (see line 41 till 48).

File: docker-compose.yml (hunk 1 above)
Comment (Taronyuu) [reply docker-compose.yml:29]: This way data is persistent for the MySQL server.

File: entrypoint.sh (hunk 2 above)
Comment (Taronyuu) [reply entrypoint.sh:12]: That is true, but this way the Dockerfile can be used for production and local development. Because when building it we want to run it, but it should also run whenever the Dockerfile is ran from the docker compose file.

**Comment by Taronyuu** [comment]
I'll reply to your points here:

> Wouldn't it make sense to move entrypoint.sh into the docker/ directory as well?

Yes you are right, we should do that.

> I agree with updating the MySQL version from 5.7 to 8, just not sure if we should include it in this scope

Fair point, I don't think there is a problem with this but we can move it out of the scope.


> The entire existence of the BUDGET_VERSION argument becomes pointless if we use COPY ., so I want to know; do you think it makes more sense to ditch it entirely and rely on the COPY command?

Yes, I think the `BUDGET_VERSION` argument doesn't hold any value. You will need to clone the repository anyway to have the Dockerfile. It is only introducing a dependency onto GitHub.

**Comment by range-of-motion** [comment]
I've committed most of these changes individually over the past few days, so I'm going to close this. If you think there are additional improvements to be made, let me know!
